brain/detail/circuit.h

```diff
@@ -220,14 +187,313 @@ class Circuit::Impl
     virtual Quaternionfs getRotations(const GIDSet& gids) const = 0;
     virtual Strings getMorphologyNames(const GIDSet& gids) const = 0;
 
-    URI getMorphologyURI(const std::string& name) const
+    virtual URI getMorphologyURI(const std::string& name) const = 0;
+
+    virtual const brion::URI& getCircuitSource() const = 0;
+    virtual void saveMorphologyToCache(
+        const std::string& uri, const std::string& hash,
+        neuron::MorphologyPtr morphology) const = 0;
+
+    virtual CachedMorphologies loadMorphologiesFromCache(
+        const std::set<std::string>& hashes) const = 0;
+    virtual void saveSynapsePositionsToCache(
+        const uint32_t gid, const std::string& hash,
+        const brion::SynapseMatrix& value) const = 0;
+    virtual CachedSynapses loadSynapsePositionsFromCache(
+        const Strings& keys) const = 0;
+    virtual void _findSynapsePositionsColumns() const = 0;
+    virtual const brion::SynapseSummary& getSynapseSummary() const = 0;
+    virtual const brion::Synapse& getSynapseAttributes(
+        const bool afferent) const = 0;
+    virtual const brion::Synapse& getAfferentProjectionAttributes(
+        const std::string& name) const = 0;
+    virtual const brion::Synapse* getSynapseExtra() const = 0;
+    virtual const brion::Synapse& getSynapsePositions(
+        const bool afferent) const = 0;
+    virtual const URI& getSynapseSource() const = 0;
+};
+
+class SonataCircuit : public Circuit::Impl
+{
+public:
+    explicit SonataCircuit(const URI& uri)
+        : config(uri)
+        , basePath(boost::filesystem::path(uri.getPath()).parent_path())
+    {
```

Comment:
The idea is to open the node group in the constructor, but you shouldn't read any datasets here. The functions that read the datasets must read at most, the slice of data comprised between the lowest and largest gids. Look into MVD3::getPositions to get an idea.